### PR TITLE
Reyna logging

### PR DESCRIPTION
**Pull request by azaka01**
Add logging to Reyna without any third party dependencies. Uses a delegate exposed through an interface to call back into the Logger object in ElemezServiceHost.
1. Create ILogDelegate with Log method signature (logLevel, logMessage, array of params)
2. Add EnableLogging(ILogDelegate) method in IReyna to allow the calling application to pass in a LoggerAdapter delegate.
3. Add ReynaLogger class for Logger API support in Reyna. Simple to use just call Logger.Error, Info etc after Initialisation.
4. Register a singleton instance of ReynaLogger in UnityHelper and expose IReynaLogger in various Reyna classes.
5. Add useful Log statements.
6. Make relevant changes to ElemezServiceHost in order to enable logging in Reyna which is part of another pull request in the x86 branch.

### CODE REVIEW

File: src/reyna/Reyna/Http/HttpClient.cs [hunk 1]
```diff
@@ -12,11 +13,13 @@ public sealed class HttpClient : IHttpClient
...
         public IConnectionManager ConnectionManager { get; set; }
         private IWebRequest webRequest;
+        private IReynaLogger Logger;
 
-        public HttpClient(IConnectionManager connectionManager, IWebRequest webRequest)
+        public HttpClient(IConnectionManager connectionManager, IWebRequest webRequest, IReynaLogger logger)
         {
             this.ConnectionManager = connectionManager;
             this.webRequest = webRequest;
+            Logger = logger;
```
Comment (b2mdevelopment) [inline src/reyna/Reyna/Http/HttpClient.cs:22]: you might need to add a test that if the logger is null to throw NullArgumentException.
@Youhana-Hana

File: src/reyna/Reyna/Http/ReynaWebRequest.cs [hunk 2]
```diff
@@ -70,6 +76,7 @@ public Result Send(string content)
             {
                 var response = webException.Response as HttpWebResponse;
                 statusCode = this.GetStatusCode(response);
+                Logger.Error("ReynaWebRequest.Send Error {0} Status code {1}", webException.ToString(), statusCode);
```
Comment (b2mdevelopment) [inline src/reyna/Reyna/Http/ReynaWebRequest.cs]: better to log exception object instead of webException.ToString() to get full stack
@Youhana-Hana

**Comment by b2mdevelopment** [comment]
Please consider above comments.

:100: 

@Youhana-Hana

File: src/reyna/Reyna/Http/HttpClient.cs (hunk 1 above)
Comment (azaka01) [inline src/reyna/Reyna/Http/HttpClient.cs:22]: Fair point, but Logger is actually ReynaLogger which is created by the UnityContainer so this will never be null. The delegate LogHandler within ReynaLogger is null to begin with so will not log by default. The LogHandler gets set when the caller enables logging ReynaService::EnableLogging so not sure if the exception should be thrown from there but it will need to be caught by the calling application. There are checks in ReynaLogger if (LogHandler != null).
